tests/asynch/basic.rs

```diff
@@ -311,3 +311,58 @@ fn drop_shuttle_yield_future() {
     assert_eq!(2, async_accesses_clone.load(Ordering::SeqCst));
     assert_eq!(1, post_yield_accesses_clone.load(Ordering::SeqCst));
 }
+
+// This test checks two behaviors. First, it checks that a future can poll another future
+// and wake itself in the same poll() call. Second, it checks that a waiter can finish
+// before the waitee task.
+#[test]
+fn wake_self_on_join_handle() {
+    use futures::{Future, FutureExt};
+    use shuttle::asynch::JoinHandle;
+    use std::pin::Pin;
+    use std::task::{Context, Poll};
+
+    check_dfs(
+        || {
+            let yielder = asynch::spawn(async move {
+                asynch::yield_now().await;
+            });
+
+            struct Timeout {
+                inner: JoinHandle<()>,
+                counter: u8,
+            }
```

Comment:
I'd write this as
```
struct Timeout<F: Future<Output = ()>> {
    inner: F,
    counter: u8,
}
```
rather than explicitly naming the JoinHandle type

Reply:
Sounds good, changing to
```
struct Timeout<F: Future> {
    inner: Pin<Box<F>>,
    counter: u8,
}
```